lib/Controller/ChatController.php

```diff
@@ -516,6 +501,45 @@ public function receiveMessages(int $lookIntoFuture,
 			];
 		}
 
+		/**
+		 * Gather information to expose $message['reactions']['self']
+		 */
+		$messageIdsWithReactions = array_map(
+			static fn (array $message) => $message['id'],
+			array_filter($messages, static fn (array $message) => !empty($message['reactions']))
+		);
+
+		$parentsWithReactions = array_map(
+			static fn (array $message) => ['parent' => $message['parent']['id'], 'message' => $message['id']],
+			array_filter($messages, static fn (array $message) => !empty($message['parent']['reactions']))
+		);
+
+		$parentMap = $parentIdsWithReactions = [];
+		foreach ($parentsWithReactions as $entry) {
+			// Create a map, so we can translate the parent's $messageId to the correct child entries
+			$parentMap[(int) $entry['parent']] ??= [];
+			$parentMap[(int) $entry['parent']][] = (int) $entry['message'];
+			$parentIdsWithReactions[] = (int) $entry['parent'];
+		}
+
+		$idsWithReactions = array_unique(array_merge($messageIdsWithReactions, $parentIdsWithReactions));
+
+		$reactionsById = $this->reactionManager->getReactionsForMessages($this->participant, $idsWithReactions);
+		foreach ($reactionsById as $messageId => $reactions) {
+			if (isset($messages[$commentIdToIndex[$messageId]])) {
+				$messages[$commentIdToIndex[$messageId]]['reactions']['self'] = $reactions;
+			}
+
+			// Add the self part also to potential parent elements
+			if (isset($parentMap[$messageId])) {
+				foreach ($parentMap[$messageId] as $mid) {
+					if (isset($messages[$commentIdToIndex[$mid]])) {
+						$messages[$commentIdToIndex[$mid]]['parent']['reactions']['self'] = $reactions;
+					}
+				}
+			}
+		}
+
```

Comment:
Will be good put this in a method and reduce the responsibility of controller. Maybe, a good place to put the new method is in ReactionManager.

args: $message, $messages
return: $messages

Reply:
I dislike this pattern of modifying local content in submethods in 3rd level code.
I'm fine with moving it to a method in the controller, but it's too unique to this place (like the parent loading just before) to put it somewhere else.

Reply:
Move is good to make more readable the code and understood the main purpose of methods.
I suggested to move to ReactionManager only to maintain reactions code in the same place, but, no problem to maintain in a private method of ChatController